src/mlpack/methods/ann/layer/subview.hpp

```diff
@@ -57,9 +61,11 @@ class Subview
   void Forward(InputType&& input, OutputType&& output)
   {
     // Check if input has been selected as required.
-    if ((input.n_rows != (end - begin + 1)) && (end != 0))
+    if (((input.n_rows != (endRow - beginRow + 1)) ||
+        (input.n_cols != (endCol - beginCol +1)))
+        && !(endRow == 0 && endCol == 0))
     {
-      output = arma::mat(&input(begin), end - begin + 1, input.n_cols, false);
+      output = input.submat(beginRow, beginCol, endRow, endCol);
```

Comment:
Thanks for looking into the issue, I guess it might be useful to combine the both ideas. So if the input is a vector ``n_cols = 1`` or if ``beginCol = 0`` and ``endCol = 0``, we can use the advanced constructor to avoid the copy. Let me know what you think.

Reply:
You are right. I think `beginCol != 0` is the only  case where submatrix is required. Seeing that advanced constructor can work for matrices too if `beginCol` is 0.

Reply:
I'm not sure if the condition is right to use the advanced constructor.  I agree that definitely if `n_cols == 1` then we can use the advanced constructor, using the memory from `&input(beginRow)` to `&input(endRow)`.  But I think the second condition should be `beginRow == 0` and `endRow == input.n_rows - 1`.

Reply:
Based on a few trials this is what I have understood: For any matrix whose subview has `beginCol == 0`, advanced constructor can be used using the memory from `&input(beginRow)` and specifying the number of rows and number of columns, i guess we are able to do this because armadillo uses column major. When `beginCol != 0` we can't access the starting index using the same method, hence `submat` was needed. Please correct me if I am mistaken. I tried out this on a few cases and it is working. Thus I couldn't understand the need of the second condition you mentioned, it would be great if you could help me with the example of a subview.

Reply:
Since Armadillo is column-major, this means that every point in every column is contiguous in memory.  So you are right that if `beginCol == 0`, then the beginning of the memory we need to use is just `&input(beginRow)` (or, alternately, `input.memptr() + beginRow`).

But let's consider the case where `beginRow == 0` and `endRow == input.n_rows - 1`.  In this situation, we are using all of the rows of the matrix.  So if we have any number of columns, they will be contiguous in memory, using the memory from `&input(beginCol * n_rows)` to `&input((endCol + 1) * n_rows - 1)`; or, alternately expressed, from `input.colptr(beginCol)` to `input.colptr(endCol) - 1`.  Does that make more sense?  Let me know if I can clarify further.

Reply:
You are right, I see what you meant now. Thanks for the clarification. I'll make the necessary edits. :)